File: python/creational/04_builder.py
"""
设计模式: 建造者模式 (Builder)
分类: 创建型

意图: 将复杂对象的构建与其表示分离，使同样的构建过程可以创建不同的表示。
核心思想: Builder 分步构建，Director 定义构建顺序，支持链式调用。
编程范式: OOP — 流式接口 (Method Chaining) + 职责分离。
"""

from __future__ import annotations
from dataclasses import dataclass, field


# ── 产品 ──────────────────────────────────────────────────────────────────
@dataclass
class Pizza:
    dough: str = ""
    sauce: str = ""
    toppings: list[str] = field(default_factory=list)

    def __str__(self) -> str:
        return f"Pizza [{self.dough} | {self.sauce} | {', '.join(self.toppings)}]"


# ── 建造者（流式接口）────────────────────────────────────────────────────
class PizzaBuilder:
    def __init__(self) -> None:
        self._pizza = Pizza()

    def set_dough(self, dough: str) -> PizzaBuilder:
        self._pizza.dough = dough
        return self

    def set_sauce(self, sauce: str) -> PizzaBuilder:
        self._pizza.sauce = sauce
        return self

    def add_topping(self, topping: str) -> PizzaBuilder:
        self._pizza.toppings.append(topping)
        return self

    def build(self) -> Pizza:
        result, self._pizza = self._pizza, Pizza()
        return result


# ── 导演 ──────────────────────────────────────────────────────────────────
class Director:
    @staticmethod
    def build_margherita(b: PizzaBuilder) -> Pizza:
        return (b.set_dough("thin")
                 .set_sauce("tomato")
                 .add_topping("mozzarella")
                 .add_topping("basil")
                 .build())

    @staticmethod
    def build_pepperoni(b: PizzaBuilder) -> Pizza:
        return (b.set_dough("thick")
                 .set_sauce("spicy")
                 .add_topping("pepperoni")
                 .add_topping("cheddar")
                 .build())


# ── 演示 ──────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    builder = PizzaBuilder()
    print(Director.build_margherita(builder))
    print(Director.build_pepperoni(builder))

File: cpp/creational/04_builder.cpp
/**
 * 设计模式: 建造者模式 (Builder)
 * 分类: 创建型
 *
 * 意图: 将一个复杂对象的构建与其表示分离，使同样的构建过程可以创建不同的表示。
 * 核心思想: Director 控制构建步骤，Builder 实现每步细节，Product 是最终结果。
 * 编程范式: OOP — 流式接口 + 职责分离。
 */

#include <iostream>
#include <string>
#include <vector>

// ── 产品 ──────────────────────────────────────────────────────────────────
struct Pizza {
    std::string              dough;
    std::string              sauce;
    std::vector<std::string> toppings;

    void print() const {
        std::cout << "Pizza [" << dough << " | " << sauce << " | ";
        for (auto& t : toppings) std::cout << t << " ";
        std::cout << "]\n";
    }
};

// ── 建造者接口 ────────────────────────────────────────────────────────────
struct PizzaBuilder {
    virtual void setDough(const std::string& d)    = 0;
    virtual void setSauce(const std::string& s)    = 0;
    virtual void addTopping(const std::string& t)  = 0;
    virtual Pizza getResult()                       = 0;
    virtual ~PizzaBuilder() = default;
};

// ── 具体建造者 ────────────────────────────────────────────────────────────
struct ConcretePizzaBuilder : PizzaBuilder {
    Pizza pizza;
    void setDough(const std::string& d)   override { pizza.dough = d; }
    void setSauce(const std::string& s)   override { pizza.sauce = s; }
    void addTopping(const std::string& t) override { pizza.toppings.push_back(t); }
    Pizza getResult()                     override { return pizza; }
};

// ── 导演 ──────────────────────────────────────────────────────────────────
struct Director {
    void buildMargherita(PizzaBuilder& b) const {
        b.setDough("thin");
        b.setSauce("tomato");
        b.addTopping("mozzarella");
        b.addTopping("basil");
    }
    void buildPepperoni(PizzaBuilder& b) const {
        b.setDough("thick");
        b.setSauce("spicy");
        b.addTopping("pepperoni");
        b.addTopping("cheddar");
    }
};

int main() {
    Director director;
    ConcretePizzaBuilder builder;

    director.buildMargherita(builder);
    builder.getResult().print();

    builder = {};  // reset
    director.buildPepperoni(builder);
    builder.getResult().print();

    return 0;
}
